java/gui/src/main/java/com/squareup/subzero/SubzeroGui.java

```diff
@@ -180,4 +203,109 @@ private void uiTest() throws Exception {
       screens.exception(e);
     }
   }
+
+  private static class IllegalTxSignTestFileFormatException extends Exception {
+    public IllegalTxSignTestFileFormatException(String msg) {
+      super(msg);
+    }
+  }
+
+  private void signTxTest() throws Exception {
+
+    // TODO: implement HSM test
+    if (nCipher) {
+      System.out.println("Transaction signing regression test not yet implemented for nCipher");
+      return;
+    }
+
+    // Passed and failed test cases, for valid and invalid test vectors.
+    // For a valid test vector, test passes (ok) if and only if subzero response matches
+    // expected response.
+    // For an invalid test vector, test passes (ok) if and only if subzero response does not
+    // match expected response.
+    int ok_valid = 0;
+    int fail_valid = 0;
+    int ok_invalid = 0;
+    int fail_invalid = 0;
+
+    // Read request & expected response from src/main/resources/txsign-testvectors/.
+    // The request and expected response are pre-generated, based64-encoded proto buffers
+    FileResourceUtils util = new FileResourceUtils();
+    List<Path> txsignRegressionTestsPath = util.getPathsFromResourcesJAR("txsign-testvectors");
+    for (Path path : txsignRegressionTestsPath) {
+      String filePathInJAR = path.toString();
+      if (filePathInJAR.startsWith("/")) {
+        filePathInJAR = filePathInJAR.substring(1, filePathInJAR.length());
+      }
+
+      try {
+        InputStreamReader inStreamReader =
+            new InputStreamReader(util.getFileFromResourceAsStream(filePathInJAR),
+                StandardCharsets.UTF_8);
+        BufferedReader reader = new BufferedReader(inStreamReader);
+
+        // Get input from file
+        String request = reader.readLine();
+        if (request == null || !request.startsWith("request:")) {
+          throw new IllegalTxSignTestFileFormatException(
+              path.toString() + ": malformed file, request not found");
+        }
+
+        // Get expected output from file
+        String expectedResponse = reader.readLine();
+        if (expectedResponse == null || !expectedResponse.startsWith("response:")) {
+          throw new IllegalTxSignTestFileFormatException(
+              filePathInJAR + ": malformed file, response not found");
+        }
+        request = request.substring("request:".length(), request.length());
+        expectedResponse =
+            expectedResponse.substring("response:".length(), expectedResponse.length());
+
+        byte[] proto = base64().decode(request);
+
+        CommandRequest commandRequest = CommandRequest.parseFrom(proto);
+
+        CommandResponse response =
+            CommandHandler.dispatch(this, new InternalCommandConnector(hostname, port),
+                commandRequest);
+
+        String encoded = base64().encode(response.toByteArray());
+
+        if (encoded.equals(expectedResponse)) {
```

Comment:
suggest simplifying logic here if possible, maybe erroring directly on mismatch?

Reply:
We clarified the cases of "valid" and "invalid" test vectors, which are corresponding to positive and negative test cases. Will keep it as is for now.